2015/presentations/Expert system in Python in 5 minutes/text.md

```diff
@@ -0,0 +1,144 @@
+
+    pip install pyclips
+
+Once you install it now we can start writing our first simple rules engine. The module fully embeds the CLIPS engine, with COOL (Clips Object Oriented Language) and environments support. This allows you to use very clear and easy to read syntax to write very complex expert systems.
+
+# Dive into..
+
+## Registering Python function
+
+We can register Python function which we later on can call from CLIPS engine whenever any of the facts match any of rules. Example:
+
+    import clips
+    
+    class RulesEngine:
+        def __init__(self):
+            clips.Conserve = True
+            clips.RegisterPythonFunction(self.myDebug, 'myDebug')
+            
+        def myDebug(self, msg):
+            print "my debug message with content: {0}" .format(msg)
+
+Now how to load rules into rules engine. Let's assume that our system has all the rules stores in one folder and rules files are named with extension .clp
+
+## Loading rules
+
+Loading such a rules files can be done like this
+
+
+    import clips
+    
+    class RulesEngine:
+        def __init__(self):
+            # registering python/clips functions hooks
+            # of course also we can have here some log objects initialized etc
+            
+        def loadClipsRules(self):
+            # rules_folder variable can be taken from ini file, etc
+            clips.BatchStar(os.path.join(rules_folder, 'Constants.clp')) # as a single file
+            
+            # or batch loading
+            for f in filter(lambda s: s.endswith('.clp'), os.listdir(rules_folder)):
+                clips.BatchStar(os.path.join(rules_folder, f))
+                # now lets check if engine did not have any errors
+                e = clips.ErrorStream.Read()
+                if e:
+                    self.log.warning('Engine ini process has errors!\n%s' % e)
+                    raise Exception()
+
+## Debug
+
+
+Now if we want to debug any of the rules that Clipe Enigne is firing or what kind of state is being updated you can dump them by adding dumping method, example:
+
+    def dumpOutput(self):
+        result = clips.TraceStream.Read()
+        if result:
+            self.log.info('CLIPS traceback:\n%s' % result)
+
+        result = clips.StdoutStream.Read()
+        if result:
+            self.log.info('CLIPS stdout:\n%s' % result)
+
+        result = clips.ErrorStream.Read()
+        if result:
+            self.log.warning('Errors:\n%s' % result)
+
+
+## Facts
+
+Before we are going to dive into details how to process facts and take some actions what first comes first - loading facts. To simplify this article let's assume that fact is a very simple state which is going to be checked by rules engine. If some states of an entity are going to match any of the rules then that particular rule(s) is going to be fired. Fired means that engine can modify any of its internal objects (another facts) or for example call one of the registered before Python functions.
```

Comment:
Please reword first sentence.
"for example call one of the registered before Python functions" -> "call one of the previously registered Python functions"